pkg/provisioner/ironic/ironic.go

```diff
@@ -1743,42 +1743,67 @@ func (p *ironicProvisioner) PowerOn() (result provisioner.Result, err error) {
 
 	if ironicNode.PowerState != powerOn {
 		if ironicNode.TargetPowerState == powerOn {
-			p.log.Info("waiting for power status to change")
-			return operationContinuing(powerRequeueDelay)
+			if ironicNode.LastError != "" {
+				p.log.Info("PowerOn operation failed", "msg", ironicNode.LastError)
+				return operationFailed(fmt.Sprintf("PowerOn operation failed: %s",
+					ironicNode.LastError))
+			} else {
+				p.log.Info("waiting for power status to change")
+				return operationContinuing(powerRequeueDelay)
+			}
 		}
-		result, err = p.changePower(ironicNode, nodes.PowerOn)
-		switch err.(type) {
-		case nil:
-		case HostLockedError:
-		default:
-			return transientError(errors.Wrap(err, "failed to power on host"))
+		if result, err = p.changePower(ironicNode, nodes.PowerOn); err == nil {
+			p.publisher("PowerOn", "Host powered on")
+			return result, nil
+		} else {
+			switch err.(type) {
+			case HostLockedError:
+				return retryAfterDelay(powerRequeueDelay)
+			default:
+				return transientError(errors.Wrap(err, "failed to PowerOn node"))
+			}
 		}
-		p.publisher("PowerOn", "Host powered on")
 	}
-
 	return result, nil
 }
 
 // PowerOff ensures the server is powered off independently of any image
 // provisioning operation.
-func (p *ironicProvisioner) PowerOff(rebootMode metal3v1alpha1.RebootMode) (result provisioner.Result, err error) {
+func (p *ironicProvisioner) PowerOff(rebootMode metal3v1alpha1.RebootMode, forceHardPowerOff bool) (result provisioner.Result, err error) {
 	p.log.Info(fmt.Sprintf("ensuring host is powered off (mode: %s)", rebootMode))
 
-	if rebootMode == metal3v1alpha1.RebootModeHard {
-		result, err = p.hardPowerOff()
-	} else {
-		result, err = p.softPowerOff()
+	softPowerOffSuccessful := true
+
+	if rebootMode != metal3v1alpha1.RebootModeHard && !forceHardPowerOff {
+		if result, err = p.softPowerOff(); err == nil {
+			return result, nil
+		} else {
+			switch err.(type) {
+			// In case of soft power off is unsupported or has failed,
+			// we activate hard power off.
```

Comment:
Honestly, I think this logic should be a new API flag on ironic. It's not BMO specific, and implementing it here needs tracking what action is in progress.

Reply:
While that API flag is being added in Ironic, how should we handle soft power off failures in BMO?